### PR TITLE
Merge compressed columns functionality from 5.6 to 5.7

### PR DESCRIPTION
**Pull request by percona-ysorokin**

### CODE REVIEW

**Comment by percona-ysorokin** [comment]
Jenkins build link (with --big-test)
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/487/
Fix for 32-bit CentOS is already included - a new Jenkins link will be added when the build is ready.

**Comment by percona-ysorokin** [comment]
An updated Jenkins build link (with --big-test)
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/490/

**Comment by laurynas-biveinis** [comment]
Is column compression disabled for internal temp tables only, or for CREATE TEMPORARY TABLE too?

File: mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result [hunk 1]
```diff
@@ -0,0 +1,1405 @@
+CREATE COMPRESSION_DICTIONARY dict1('foo');
+CREATE COMPRESSION_DICTIONARY dict2('bar');
+# from 'BLOB' to 'xxx'
+*** "a BLOB" -> "a BLOB", ALGOTIRHM=DEFAULT
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result]: "ALGOTIRHM"

File: mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result [hunk 2]
```diff
@@ -0,0 +1,1405 @@
...
+original_dictionary_reference_does_not_exist
+1
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result:6]: This result file could be shrinked by assert.inc use? (for 5.6 too; not for this PR)

File: mysql-test/suite/innodb/r/xtradb_compressed_columns_mysqldump.result [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+) ENGINE=InnoDB DEFAULT CHARSET=latin1
+SET @short_value = REPEAT(@dictionary_word, 2);
+SET @long_value = REPEAT(@dictionary_word, @@innodb_compressed_columns_threshold / LENGTH(@dictionary_word) + 2);
+INSERT INTO t1(a, b, c, d) VALUES (@short_value, @short_value, @short_value, @short_value);
+INSERT INTO t1(a, b, c, d) VALUES (@long_value, @long_value, @long_value, @long_value);
+INSERT INTO t2(a, b, c, d) VALUES (@short_value, @short_value, @short_value, @short_value);
+INSERT INTO t2(a, b, c, d) VALUES (@long_value, @long_value, @long_value, @long_value);
+# Importing SQL dump without compressed column extensions
+number_of_short_rows_in_first_match
+1
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/r/xtradb_compressed_columns_mysqldump.result:74]: Likewise

File: mysql-test/suite/innodb/r/xtradb_compressed_columns_with_dictionaries.result [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+id INT,
+a BLOB COLUMN_FORMAT COMPRESSED,
+b BLOB COLUMN_FORMAT COMPRESSED WITH COMPRESSION_DICTIONARY d1
+) ENGINE=InnoDB;
+SELECT table_id INTO @table_id FROM information_schema.innodb_sys_tables
+WHERE name=CONCAT('db1', '/', 't1');
+SELECT COUNT(*) = 1 AS only_one_dictionary_reference FROM information_schema.xtradb_zip_dict_cols
+WHERE table_id = @table_id;
+only_one_dictionary_reference
+1
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/r/xtradb_compressed_columns_with_dictionaries.result:152]: Likewise

File: sql/sql_lex.cc [hunk 5]
```diff
@@ -53,6 +53,8 @@ sys_var *trg_new_row_fake_var= (sys_var*) 0x01;
 */
 const LEX_STRING null_lex_str= {NULL, 0};
 const LEX_STRING empty_lex_str= {(char *) "", 0};
+const LEX_CSTRING null_lex_cstr= {NULL, 0};
```
Comment (laurynas-biveinis) [inline sql/sql_lex.cc:56]: Duplication

File: sql/sql_lex.h [hunk 6]
```diff
@@ -1788,6 +1789,8 @@ struct st_sp_chistics
 
 extern const LEX_STRING null_lex_str;
 extern const LEX_STRING empty_lex_str;
+extern const LEX_CSTRING null_lex_cstr;
```
Comment (laurynas-biveinis) [inline sql/sql_lex.h:1792]: Duplication

File: sql/sql_table.cc [hunk 7]
```diff
@@ -3592,6 +3593,28 @@ mysql_prepare_create_table(THD *thd, HA_CREATE_INFO *create_info,
...
     const CHARSET_INFO *save_cs;
 
+    /* check if the column is compressible.*/
+    if (sql_field->sql_type == MYSQL_TYPE_TINY_BLOB ||
+        sql_field->sql_type == MYSQL_TYPE_MEDIUM_BLOB ||
+        sql_field->sql_type == MYSQL_TYPE_BLOB ||
+        sql_field->sql_type == MYSQL_TYPE_LONG_BLOB ||
+        sql_field->sql_type == MYSQL_TYPE_VARCHAR)
+    {
+      DBUG_EXECUTE_IF("enforce_all_compressed_columns",
```
Comment (laurynas-biveinis) [inline sql/sql_table.cc:3609]: Please do a enforce_all_compressed_columns / remove_compressed_attributes_for_keys run if you haven't

File: storage/innobase/handler/ha_innodb.cc [hunk 8]
```diff
@@ -5727,6 +5836,9 @@ innobase_vcol_build_templ(
         templ->mbminlen = dict_col_get_mbminlen(col);
         templ->mbmaxlen = dict_col_get_mbmaxlen(col);
         templ->is_unsigned = col->prtype & DATA_UNSIGNED;
+	templ->compressed = (field->column_format()
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc]: Is it surrounding or patch indentation which is wrong?

File: storage/innobase/handler/ha_innodb.cc [hunk 9]
```diff
@@ -8171,8 +8372,11 @@ calc_row_difference(
 		case DATA_POINT:
 		case DATA_VAR_POINT:
 		case DATA_GEOMETRY:
-			o_ptr = row_mysql_read_blob_ref(&o_len, o_ptr, o_len);
-			n_ptr = row_mysql_read_blob_ref(&n_len, n_ptr, n_len);
+			/* Do not compress blob column while comparing*/
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc]: Space after "comparing"

**Comment by percona-ysorokin** [comment]
Yes, for internal temp tables only.
`CREATE TEMPORARY TABLE` behavior wasn't changed

``` sql
CREATE TEMPORARY TABLE t1(id INT, a BLOB COLUMN_FORMAT COMPRESSED) ENGINE=InnoDB;
SHOW CREATE TABLE t1;
```

```
Table   Create Table
t1      CREATE TEMPORARY TABLE `t1` (
  `id` int(11) DEFAULT NULL,
  `a` blob /*!50633 COLUMN_FORMAT COMPRESSED */
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

File: storage/innobase/handler/ha_innodb.h [hunk 10]
```diff
@@ -351,6 +351,15 @@ class ha_innobase: public handler
...
 
+	/** This function reads zip dict-related info from SYS_ZIP_DICT
+	and SYS_ZIP_DICT_COLS for all columns marked with
+	COLUMN_FORMAT_TYPE_COMPRESSED flag and updates
+	zip_dict_name / zip_dict_data for those which have associated
+	compression dictionaries.
+	*/
+	virtual void update_field_defs_with_zip_dict_info();
+
+
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.h:365]: One less empty line

File: storage/innobase/handler/ha_innodb.h [hunk 11]
```diff
@@ -1108,3 +1117,34 @@ innobase_build_index_translation(
 	INNOBASE_SHARE*		share);	/*!< in/out: share structure
 					where index translation table
 					will be constructed in. */
+
+/** Compression dictionary id container */
+typedef std::map<uint16, ulint, std::less<uint16>,
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.h:1125]: typedefs for uint16 and ulint?

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_consistency_antelope_compact.test [hunk 12]
```diff
@@ -0,0 +1,7 @@
+--source include/big_test.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns_consistency_antelope_compact.test:1]: Should these be marked as big in 5.6 too?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: sql/sql_lex.cc (hunk 5 above)
Comment (percona-ysorokin) [reply sql/sql_lex.cc:56]: LEX_**C**STRING (const char *)

File: mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result (hunk 1 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb/r/xtradb_compressed_columns_alter_table.result]: fixed

File: storage/innobase/handler/ha_innodb.cc (hunk 8 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/ha_innodb.cc]: fixed (although previous lines use spaces for indentation)

File: storage/innobase/handler/ha_innodb.cc (hunk 9 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/ha_innodb.cc]: fixed

File: storage/innobase/handler/ha_innodb.h (hunk 10 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/ha_innodb.h:365]: fixed

File: storage/innobase/handler/ha_innodb.h (hunk 11 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/ha_innodb.h:1125]: `uint16` is exactly how `Field::field_index` is defined

``` c++
uint16        field_index;            // field number in fields array
```

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_consistency_antelope_compact.test (hunk 12 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns_consistency_antelope_compact.test:1]: on my list

**Comment by percona-ysorokin** [comment]
Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/517/
(Yakkety included)

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test [hunk 13]
```diff
@@ -0,0 +1,204 @@
+--source include/have_innodb.inc
+
+# removing STRICT_TRANS_TABLES from sql_mode
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: Why?

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_virtual.test [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+  a BLOB COLUMN_FORMAT COMPRESSED WITH COMPRESSION_DICTIONARY numbers,
+  c BLOB GENERATED ALWAYS AS (a) VIRTUAL,
+  INDEX(c(100))
+) ENGINE=InnoDB;
+
+INSERT INTO t VALUES (REPEAT('a', 16000), DEFAULT);
+
+SET GLOBAL debug="+d,ib_purge_virtual_index_callback";
+UPDATE t SET a = REPEAT('m', 16000) WHERE a LIKE "aaa%";
+SELECT SLEEP(3);
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns_virtual.test]: 1) SLEEP is a testcase smell - while here random scheduling will not cause a testcase failure, that would require clarification; 2) it is not safe to update global debug variable while background threads may access it (purge in this instance). Can this be replaced by e.g.

wait for purge to complete;
SET GLOBAL debug="+d,ib_purge_virtual_index_callback";
UPDATE ...
wait for purge to complete;
SET GLOBAL debug="-d,ib_purge_virtual_index_callback";

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_with_dictionaries.test [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+# dropping table with references to compression dictionaries must also clean
+# SYS_ZIP_DICT_COLS InnoDB system table
+--eval DROP TABLE $table_name
+
+SELECT COUNT(*) = 1 AS compression_dictionary_exists FROM information_schema.xtradb_zip_dict
+  WHERE id = @dict_id;
+SELECT COUNT(*) = 0 AS references_cleared FROM information_schema.xtradb_zip_dict_cols
+  WHERE table_id = @table_id;
+
+# check if tables deleted implicitly via "DROP DATABASE" aslo perform
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns_with_dictionaries.test]: "also"

File: sql/field.cc [hunk 16]
```diff
@@ -7988,8 +8016,13 @@ Field *Field_varstring::new_key_field(MEM_ROOT *root,
 
 uint Field_varstring::is_equal(Create_field *new_field)
 {
+  /*
+    changing column format to/from compressed or changing assiciated
```
Comment (laurynas-biveinis) [inline sql/field.cc]: associated

File: storage/innobase/handler/handler0alter.cc [hunk 17]
```diff
@@ -1544,6 +1544,15 @@ innobase_col_to_mysql(
 		field->reset();
 
 		if (field->type() == MYSQL_TYPE_VARCHAR) {
+			if (field->column_format() ==
+				COLUMN_FORMAT_TYPE_COMPRESSED) {
+				/* Skip compressed varchar column when
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/handler0alter.cc]: Why?

File: storage/innobase/handler/handler0alter.cc [hunk 18]
```diff
@@ -2892,7 +2905,8 @@ innobase_build_col_map(
 	const dict_table_t*	new_table,
 	const dict_table_t*	old_table,
 	dtuple_t*		add_cols,
-	mem_heap_t*		heap)
+	mem_heap_t*		heap,
+	row_prebuilt_t*	prebuilt)
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/handler0alter.cc]: Indentation

File: storage/innobase/row/row0mysql.cc [hunk 19]
```diff
@@ -3666,6 +4209,12 @@ row_mysql_table_id_reassign(
 	pars_info_add_ull_literal(info, "old_id", table->id);
 	pars_info_add_ull_literal(info, "new_id", *new_id);
 
+	/* As micro-SQL does not support int4 == int8 comparisons,
```
Comment (laurynas-biveinis) [inline storage/innobase/row/row0mysql.cc:4212]: You found its name!

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_with_dictionaries.test (hunk 15 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns_with_dictionaries.test]: fixed

File: sql/field.cc (hunk 16 above)
Comment (percona-ysorokin) [reply sql/field.cc]: fixed

File: storage/innobase/handler/handler0alter.cc (hunk 18 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/handler0alter.cc]: fixed

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test (hunk 13 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: Some checks rely on data truncation being treated as a warning.
Without modifying sql_mode some INSERT statements will generate errors.

Comment (laurynas-biveinis) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: Can INSERT IGNORE be used for those (I assume only a few) statements?

Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: Good Idea.
Reworked with "INSERT IGNORE"

File: storage/innobase/handler/handler0alter.cc (hunk 17 above)
Comment (percona-ysorokin) [reply storage/innobase/handler/handler0alter.cc]: It looks like an artifact from the original "compressed columns" patch by Weixiang Zhai.
This code path is never reached.
Fragment removed.

File: mysql-test/suite/innodb/t/xtradb_compressed_columns_virtual.test (hunk 14 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns_virtual.test]: Rewritten with `--source include/wait_innodb_all_purged.inc`

**Comment by percona-ysorokin** [comment]
Jenkins build for commit d142eb0
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/529/
(includes --big-test)

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**